src/core/DY.Core.Trace.Base.fst

```diff
@@ -855,14 +814,79 @@ val trace_find_first: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label
 let trace_find_first tr e =
   Some?.v (trace_search_first tr (trace_entry_equiv e))
 
-[@@"opaque_to_smt"]
-val trace_find_last: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label_t{entry_exists tr e} ->
+val trace_find_last: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label_t ->
     Pure timestamp
-      (requires True)
+      (requires equiv_entry_exists tr e)
       (ensures fun ts ->
         ts `on_trace` tr /\
         trace_entry_equiv e (get_entry_at tr ts) /\
         forall ts'. (ts' `on_trace` tr /\ trace_entry_equiv e (get_entry_at tr ts')) ==> ts' <= ts
       )
 let trace_find_last tr e =
   Some?.v (trace_search_last tr (trace_entry_equiv e))
+
+val opt_less_defined: #a:eqtype -> option a -> option a -> bool
+let opt_less_defined o1 o2 =
+  o1 = None || o1 = o2
+
+val trace_search_first_later:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  p:(trace_entry_ label_t -> bool) ->
+  Lemma
+  (requires tr1 <$ tr2)
+  (ensures opt_less_defined (trace_search_first tr1 p) (trace_search_first tr2 p))
+  [SMTPat (tr1 <$ tr2);
+   SMTPat (trace_search_first tr1 p);
+   SMTPat (trace_search_first tr2 p);
+  ]
+let trace_search_first_later tr1 tr2 p = ()
+
+val trace_find_first_later:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  e:trace_entry_ label_t ->
+  Lemma
+  (requires
+    equiv_entry_exists tr1 e /\
+    tr1 <$ tr2
+  )
+  (ensures trace_find_first tr1 e == trace_find_first tr2 e)
+  [SMTPat (tr1 <$ tr2);
+   SMTPat (trace_find_first tr1 e);
+   SMTPat (trace_find_first tr2 e);
+  ]
+let trace_find_first_later tr1 tr2 e = ()
+
+
+/// Find the timestamp at which an event has been triggered.
+/// (Can be useful when debugging a proof.)
+
+[@@ "opaque_to_smt"]
+val find_event_triggered_at_timestamp:
+  #label_t:Type ->
+  tr:trace_ label_t -> prin:principal -> tag:string -> content:bytes ->
+  Pure timestamp
+  (requires event_triggered tr prin tag content)
+  (ensures fun i ->
+    event_triggered_at tr i prin tag content /\
+    ~(event_triggered (prefix tr i) prin tag content)
+  )
+let find_event_triggered_at_timestamp #label_t tr prin tag content =
+  trace_find_first tr (Event prin tag content)
+
+val find_event_triggered_at_timestamp_later:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  prin:principal -> tag:string -> content:bytes ->
+  Lemma
+  (requires
+    event_triggered tr1 prin tag content /\
+    tr1 <$ tr2
+  )
+  (ensures find_event_triggered_at_timestamp tr1 prin tag content == find_event_triggered_at_timestamp tr2 prin tag content)
+  [SMTPat (find_event_triggered_at_timestamp tr1 prin tag content);
+   SMTPat (find_event_triggered_at_timestamp tr2 prin tag content);
+   SMTPat (tr1 <$ tr2)
+  ]
+let find_event_triggered_at_timestamp_later #label_t tr1 tr2 prin tag content = ()
```

Comment:
It is suspicious that the proof works without revealing the definition of `find_event_triggered_at_timestamp`… is it really opaque to SMT?

Reply:
Ah, I think it is deduced from its post-condition!

Reply:
Yes, I was also surprised by this at first, but the postconditions of the search/find functions are surprisingly strong (or rather, the SMT solver is surprisingly good at making use of them).